sql/dd/impl/bootstrap/bootstrapper.cc

```diff
@@ -82,11 +82,30 @@ namespace {
 // Initialize recovery in the DDSE.
 bool DDSE_dict_recover(THD *thd, dict_recovery_mode_t dict_recovery_mode,
                        uint version) {
-  handlerton *ddse = get_dd_engine(thd);
-  if (ddse->dict_recover == nullptr) return true;
+  handlerton *ddse = nullptr;
 
+  if (opt_initialize) {
+    ddse = get_dd_engine(thd);
+  } else {
+    ddse = ha_resolve_by_legacy_type(
+        thd, bootstrap::DD_bootstrap_ctx::instance().get_actual_dd_engine());
+  }
```

Comment:
`DDSE_dict_recover`, `DDSE_dict_init`, and `sync_meta_data` use the same code, please extract a helper that returns `ddse`